rrs/views.py

```diff
@@ -0,0 +1,1005 @@
+                        FROM rrs_recipemaintainer AS rema
+                        INNER JOIN rrs_maintainer AS ma
+                        ON rema.maintainer_id = ma.id
+                        WHERE rema.history_id = %s 
+                        AND ma.name = %s;
+                    """, [date_id, maintainer])
+
+        for re in cur.fetchall():
+            recipes.append(re[0])
+        return recipes
+
+    @staticmethod
+    def get_ma_by_recipes_and_date(recipes_id, date_id=None):
+        """ Get Maintainer based on Recipes and Recipe Upstream History """
+        stats = []
+
+        if date_id:
+            qry = """SELECT rema.recipesymbol_id, ma.name
+                    FROM rrs_recipemaintainer AS rema
+                    INNER JOIN rrs_maintainer AS ma
+                    ON rema.maintainer_id = ma.id
+                    WHERE rema.history_id = %s
+                    AND rema.recipesymbol_id IN %s;"""
+            cur = connection.cursor()
+            cur.execute(qry, [str(date_id), tuple(recipes_id)])
+            stats = Raw.dictfetchall(cur)
+
+        return stats
+
+    @staticmethod
+    def get_reup_statistics(maintplan, date, date_id):
+        """ Special case to get recipes statistics removing gcc-source duplicates """
+        recipes = []
+        updated = 0
+        not_updated = 0
+        cant = 0
+        unknown = 0
+
+        for maintplanlayer in maintplan.maintenanceplanlayerbranch_set.all():
+            layerbranch = maintplanlayer.layerbranch
+            layerbranch_recipes = Raw.get_reupg_by_date(layerbranch.id, date)
+            for re in layerbranch_recipes:
+                recipes.append(re["id"])
+
+        if date_id and recipes:
+            qry = """SELECT id, status, no_update_reason
+                    FROM rrs_recipeupstream
+                    WHERE history_id = %s
+                    AND recipesymbol_id IN %s;"""
+            cur = connection.cursor()
+            cur.execute(qry, [str(date_id.id), tuple(recipes)])
+
+            for re in Raw.dictfetchall(cur):
+                if re["status"] == "Y":
+                    updated += 1
+                elif re["status"]  == "N" and re["no_update_reason"] == "":
+                    not_updated += 1
+                elif re["status"] == "N":
+                    cant += 1
+                # We count downgrade as unknown
+                else:
+                    unknown += 1
+
+        return (updated, not_updated, cant, unknown)
+
+    @staticmethod
+    def get_reup_by_recipes_and_date(recipes_id, date_id=None):
+        """ Get Recipe Upstream based on Recipes and Recipe Upstream History """
+        stats = []
+
+        if date_id:
+            qry = """SELECT recipesymbol_id, status, no_update_reason, version
+                    FROM rrs_recipeupstream
+                    WHERE history_id = %s
+                    AND recipesymbol_id IN %s;"""
+            cur = connection.cursor()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *cur*. The resource is allocated by call *connection.cursor*. Execution paths that do not contain closure statements were detected. To prevent this resource leak, close *cur* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).